feat: fix AVM circuit size

lucasxia01 opened the pull request:
Remove circuit size and num_public_inputs from AVM. Use the fixed AVM size everywhere.

jeanmon commented on barretenberg/cpp/src/barretenberg/vm2/constraining/prover.cpp:
```diff
@@ -36,7 +37,7 @@ AvmProver::AvmProver(std::shared_ptr<Flavor::ProvingKey> input_key, const PCSCom
  */
 void AvmProver::execute_preamble_round()
 {
-    const auto circuit_size = static_cast<uint32_t>(key->circuit_size);
```
I would favor that we inherit the circuit_size from the key.
The rationale is that VerifierKey and ProvingKey is the place where we pass the constant and everywhere else we inherit. Otherwise, if we use the constant MAX_AVM_TRACE_SIZE everywhere in the code, there is no need to keep a circuit_size member in ProvingKey/VerifierKey.

Asking @fcarreiro opinion on this.

jeanmon replied:
My comment applies to the rest of this PR but let us wait for @fcarreiro feedback.

fcarreiro replied:
I hadn't seen this comment. I guess I have no strong preference. However as Jean says

>  if we use the constant MAX_AVM_TRACE_SIZE everywhere in the code, there is no need to keep a circuit_size member in ProvingKey/VerifierKey.

jeanmon reviewed (commented):
Thanks for this cleaning work.

iakovenkos commented on barretenberg/cpp/src/barretenberg/vm2/constraining/prover.cpp:
```diff
@@ -100,19 +101,19 @@ void AvmProver::execute_relation_check_rounds()
     }
 
-    Sumcheck sumcheck(key->circuit_size,
+    Sumcheck sumcheck(MAX_AVM_TRACE_SIZE,
```
after my extension-by-zero PR, the prover can keep using the genuine circuit size here and in ShpleminiProver

iakovenkos replied:
I had an impression that skipping is highly optimized in AVM, but using genuine size allows to skip the entire rounds from `log_circuit_size` to `MAX_AVM_TRACE_SIZE_LOG_N - 1` and for example don't compute/allocate `GateSeparator` of size 2^21.

lucasxia01 replied:
as discussed, not relevant if we just fix the AVM size

iakovenkos commented on barretenberg/cpp/src/barretenberg/vm2/constraining/polynomials.cpp:
```diff
@@ -29,7 +29,7 @@ AvmProver::ProverPolynomials compute_polynomials(tracegen::TraceContainer& trace
                            poly = AvmProver::Polynomial(
                                /*memory size*/ allocated_size,
-                               /*largest possible index*/ CIRCUIT_SUBGROUP_SIZE,
+                               /*largest possible index*/ MAX_AVM_TRACE_SIZE,
```
I don't know if it affects the efficiency/memory, but prob would make sense to use the genuine dyadic circuit size

lucasxia01 replied:
yeah, we should consider using the real circuit size stuff in the AVM

iakovenkos commented on barretenberg/cpp/src/barretenberg/vm2/constraining/prover.cpp:
```diff
@@ -68,7 +69,7 @@ void AvmProver::execute_log_derivative_inverse_round()
             AVM_TRACK_TIME(std::string("prove/log_derivative_inverse_round/") + std::string(Relation::NAME),
                            (compute_logderivative_inverse<FF, Relation>(
-                               prover_polynomials, relation_parameters, key->circuit_size)));
+                               prover_polynomials, relation_parameters, MAX_AVM_TRACE_SIZE)));
```
I'd try the genuine size here as well

lucasxia01 replied:
as discussed, not relevant if we just fix the AVM size

lucasxia01 commented on barretenberg/cpp/src/barretenberg/dsl/acir_format/avm2_recursion_constraint.cpp:
```diff
@@ -70,30 +69,12 @@ void create_dummy_vkey_and_proof(Builder& builder,
     for (size_t i = 0; i < Flavor::NUM_PRECOMPUTED_ENTITIES; ++i) {
         set_dummy_commitment(key_fields, offset);
     }
+    info("key length: ", key_fields.size());
```
should delete

lucasxia01 commented on yarn-project/bb-prover/src/prover/server/bb_prover.ts:
```diff
@@ -776,7 +771,7 @@ export class BBNativeRollupProver implements ServerCircuitProver {
     );
 
-    const proof = new Proof(rawProofBuffer, vkData.numPublicInputs);
+    const proof = new Proof(rawProofBuffer, 0);
```
I guess this was causing some serialization/deserialization issue as it tried to read a very large number of public inputs or something

fcarreiro replied:
Could you do
```
 new Proof(rawProofBuffer, /*numPublicInputs=*/0);
```
?

fcarreiro commented on barretenberg/cpp/src/barretenberg/stdlib_circuit_builders/circuit_builder_base_impl.hpp:
```diff
@@ -237,7 +237,9 @@ template <typename FF_> void CircuitBuilderBase<FF_>::failure(std::string msg)
         // We have a builder failure when we have real witnesses which is a mistake.
-        info("(Experimental) WARNING: Builder failure when we have real witnesses!"); // not a catch-all error
+        info("(Experimental) WARNING: Builder failure when we have real witnesses! Ignore if writing vk."); // not a
+                                                                                                            // catch-all
```
Weird formatting of the comment

fcarreiro commented on barretenberg/cpp/src/barretenberg/vm2/constraining/flavor.hpp:
```diff
@@ -219,18 +220,13 @@ class AvmFlavor {
 
-        ProvingKey() = default;
-        ProvingKey(const size_t circuit_size, const size_t num_public_inputs);
+        size_t circuit_size = MAX_AVM_TRACE_SIZE;
+        size_t log_circuit_size = MAX_AVM_TRACE_LOG_SIZE;
```
It seems we don't really use this in the AVM code, are they required by some other bb part?

lucasxia01 replied:
good point, I can try using these in the code

lucasxia01 replied:
no, they're not, but I made it so that the avm now uses them :)

fcarreiro commented on barretenberg/cpp/src/barretenberg/vm2/constraining/recursion/recursive_verifier.cpp:
```diff
@@ -112,15 +112,20 @@ AvmRecursiveVerifier::PairingPoints AvmRecursiveVerifier::verify_proof(
     }
 
-    const std::vector<FF> padding_indicator_array(CONST_PROOF_SIZE_LOG_N, 1);
+    // unconstrained
```
Is this bad? Is there a TODO/issue so that you guys will come back to constrain it? (especially given that you are leaving)

lucasxia01 replied:
hm, not sure why this is here lol

lucasxia01 replied:
it should be constrained because we fix the witness to a constant 1

fcarreiro commented on yarn-project/bb-prover/src/prover/server/bb_prover.ts:
```diff
@@ -589,12 +585,12 @@ export class BBNativeRollupProver implements ServerCircuitProver {
           inputSize: input.serializeWithMessagePack().length,
-          circuitSize: verificationKey.circuitSize, // FIX: wrong in VK
-          numPublicInputs: verificationKey.numPublicInputs, // FIX: wrong in VK
+          circuitSize: 1 << 21, // FIX: wrong in VK
+          numPublicInputs: 0, // FIX: wrong in VK
```
Pls remove comments.

fcarreiro reviewed (approved):
Looks ok in general, left a few comments.

Commit:
feat: fix AVM circuit size

Remove circuit size and num_public_inputs from AVM. Use the fixed AVM size everywhere.

Co-authored-by: lucasxia01 <lucasxia01@gmail.com>
